rust_src/src/dispnew.rs

```diff
@@ -136,4 +143,65 @@ pub fn internal_show_cursor_p(window: LispWindowOrSelected) -> bool {
     !win.cursor_off_p()
 }
 
+/***********************************************************************
+	   Terminal Related Lisp Functions
+***********************************************************************/
+
+/// Beep, or flash the screen.
+/// Also, unless an argument is given,
+/// terminate any keyboard macro currently executing.
+#[lisp_fn(min = "0")]
+pub fn ding(arg: Option<LispObject>) {
```

Comment:
I do not see the gain in Option<LispObject>. If the user provides no arguments then arg will be Qnil.

Reply:
My general thinking is that it's good to give the type checker as much logic as possible. In this case we take one argument, and it's a `LispObject`. But we know at compile time that its value doesn't matter at all provided that it's non-nil. That is, all that matters is whether it's nil or not. Thus we can tell the compiler to look for just two possible (sets of) values. We can't use `bool`, because any non-nil value is a valid argument, eg `(ding "ding")`, and that can't restricted. But `Option` is another way of representing two possibilities, so `Option<LispObject>` works.

Actually, even that's too broad. It also doesn't matter if the thing is a `LispObject` or not, since its value is never used. Really what it should be is something like `Option<T> where LispObject: From<T>`. Unfortunately it doesn't seem to be possible to use `<T>` with `lisp_fn` at the moment.

This isn't important for `ding` specifically of course, but it's something to think about generally, since this pattern comes up a lot.

Reply:
I see that view. I largely share it. However, we are implementing the runtime for a typeless system. And at some point we will care about efficiency. `LispObject` is an `Option` type already. It is nil or a value. I hope as we become independent of the C that `LispObject` can be a newtype for `Option<LispValue>` and nil becomes `None`. Until then, we should be wise with what we have.